### ISSUE TITLE
Provide new computed metrics for MicroProfile Metrics

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#25813]
## Description

**problem**
```
    today we provide metrics for servlets and REST requests (see https://openliberty.io/docs/latest/metrics-list.html#metrics-table)
        mpMetrics-4.0 and below
            vendor_servlet_responseTime_total_seconds{servlet="<servletName>"} 
            vendor_servlet_request_total{servlet="<servletName>"}

            base_REST_request_total{class="<fully_qualified_class_name>",method="<method_signature>"}
            base_REST_request_elapsedTime_seconds{class="<fully_qualified_class_name>",method="<method_signature>"}

        mpMetrics-5.0 and above
            servlet_responseTime_total_seconds{mp_scope="vendor",servlet="<servletName>"} 
            servlet_request_total{mp_scope="vendor",servlet="<servletName>"}

            REST_request_seconds_sum{class="<fully_qualified_class_name>",method="<method_signature>",mp_scope="base"}
            REST_request_seconds_count{class="<fully_qualified_class_name>",method="<method_signature>",mp_scope="base"}

    For example, if we have 3 servlets ("servletA", "servletB", "servletC") we would expose the following metrics in our /metrics output using mpMetrics-4.0:

            vendor_servlet_responseTime_total_seconds{servlet="servletA"} 
            vendor_servlet_request_total{servlet="servletA"}

            vendor_servlet_responseTime_total_seconds{servlet="servletB"} 
            vendor_servlet_request_total{servlet="servletB"}

            vendor_servlet_responseTime_total_seconds{servlet="servletC"} 
            vendor_servlet_request_total{servlet="servletC"}

    we use these metrics in our grafana dashboards to show the time per servlet request or per REST request, computed as follows...
        rate(vendor_servlet_responseTime_total_seconds{instance=~`[[instance]]`}[[dur]])/rate(vendor_servlet_request_total{instance=~`[[instance]]`}[[dur]])

        rate(base_REST_request_elapsedTime_seconds{instance=~`[[instance]]`}[[dur]]) / rate(base_REST_request_total{instance=~`[[instance]]`}[[dur]])

    using the above example, this would result in a timeseries for each servlet name or REST class/method combination, and works because tools like Prometheus support dividing an array of timeseries by another array of timeseries.

    the problem is, not all tools support that kind of timeseries math. eg. CloudWatch doesn't support expressions that try to do division between arrays of time series.
```

**proposed solution**
```
    we need OL to provide metrics that have already done the "time per request" computation. These new metrics should have the same labels as the metrics listed above. all new metrics must be computed using RECENT data (ewma?)

        mpMetrics-4.0 and below
            vendor_servlet_responseTime_per_request{servlet="<servletName>"}
                vendor_servlet_responseTime_total_seconds{servlet="<servletName>"} /
                vendor_servlet_request_total{servlet="<servletName>"} 

            vendor_REST_responseTime_per_request{class="<fully_qualified_class_name>",method="<method_signature>"}
                base_REST_request_total{class="<fully_qualified_class_name>",method="<method_signature>"} /
                base_REST_request_elapsedTime_seconds{class="<fully_qualified_class_name>",method="<method_signature>"}

        mpMetrics-5.0 and above
            servlet_responseTime_per_request{mp_scope="vendor", servlet="<servletName>"}
                servlet_responseTime_total_seconds{mp_scope="vendor",servlet="<servletName>"} /
                servlet_request_total{mp_scope="vendor",servlet="<servletName>"}

            vendor_REST_responseTime_per_request{class="<fully_qualified_class_name>",method="<method_signature>",mp_scope="vendor"}
                REST_request_seconds_sum{class="<fully_qualified_class_name>",method="<method_signature>",mp_scope="base"} /
                REST_request_seconds_count{class="<fully_qualified_class_name>",method="<method_signature>",mp_scope="base"}
```

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - N/A
- UFO: https://ibm.box.com/s/34dh46m7rpsv9f2q5rddkch74rba47ec
- FTS: https://github.com/OpenLiberty/open-liberty/issues/26536
- Beta Blog: https://github.com/OpenLiberty/open-liberty/issues/26406
- GA Blog: https://github.com/OpenLiberty/open-liberty/issues/26785

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [x] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [x] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [x] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [x] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [x] POC Design / UFO review requested.
  - Owner adds label `Design Review Request`
- [x] POC Design / UFO review scheduled.
  - Follow the instructions in POC-Forum repo
- [x] POC Design / UFO review completed.
- [x] POC / UFO Review follow-ons completed.
- [x] POC Design / UFO approval requested.
  - Owner adds label `Design Approval Request`
- [x] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [x] No Design requested.
  - Owner adds label `No Design Approval Request`
- [x] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`
- [x] Feature / Capability stabilization or discontinuation or N/A
  - Owner adds label `Product Management Approval Request` and notifies [OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management)
  - Approver adds label `Product Management Approved` ([OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management))
  - Note: For stabilized, superseded, and discontinued feature/capability, skip the [Beta](#beta) section of the template (you may delete it).  Otherwise, proceed as normal.

### **FAT Documentation**
- [x] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [x] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [x] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [x] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [x] PII updates are merged, or N/A. Note timing with translation shipments.

### **Innovation** (Complete 1 week before Feature Complete Date)
- [x] Consider whether any aspects of the feature may be patentable. If any identified, disclosures have been submitted.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [x] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [x] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [x] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [beta release blog post](https://openliberty.io/blog/?search=beta&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [x] Feature implementation and tests completed.
  - [x] All PRs are merged.
  - [x] All epic and child issues are closed.
  - [x] All stop ship issues are completed.
- [x] Legal: all necessary approvals granted.
- [x] Translation: All messages translated or sent for translation for upcoming release 
- [x] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [x] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [x] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Add comment `@OpenLiberty/demo-approvers Demo scheduled for EOI [Iteration Number]` to this issue.
  - Approver adds label `focalApproved:demo`.
- [x] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.
- [x] **Globalization** Translation and TVT are complete or N/A. ([OpenLiberty/globalization-approvers](https://github.com/orgs/OpenLiberty/teams/globalization-approvers))
  - Approver adds label `focalApproved:globalization`.

### **Design Approved Features**
- [x] **Accessibility** Accessibility testing completed or N/A. ([OpenLiberty/accessibility-approvers](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
  - Approver adds label `focalApproved:accessibility`.
- [x] **ID** Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **Performance** Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [x] **Serviceability** Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [x] **STE** Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [x] **SVT** System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [x] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [x] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [release blog post](https://openliberty.io/blog/?search=release&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

### **Post GA**
- [x] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([OpenLiberty/guide-assessment](https://github.com/orgs/OpenLiberty/teams/guide-assessment/members))
- [ ] **Dev Experience** Developer Experience & Tools work is complete or N/A. ([OpenLiberty/dev-experience-assessment](https://github.com/orgs/OpenLiberty/teams/dev-experience-assessment/members))
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by donbourne** [comment]
While it wouldn't make sense to do an EWMA on the total request count or total time timeseries (since they are monotonically increasing), we should be able to compute the difference in each of those values, for periodic samples, and use those differences to know the average request duration for that period, which we could then add to an EWMA to provide a reasonable estimate of the recent response time.

**Comment by donbourne** [comment]
The following is an updated list of metrics we need, based on our Grafana dashboard content:

T0 represents the time of our most recent sample of a metric, in seconds
T-1 represents the time of the previous sample of a metric, in seconds


## Process CPU Utilization Percent
Grafana dashboard (Process CPU Time): 
```
rate(base_cpu_processCpuTime_seconds{instance=~`[[instance]]`}[[cpuDur]]) / base_cpu_availableProcessors
```
Built-in Liberty metric:
```
Gauge
vendor_cpu_processCpuUtilization_percent
EWMA(((T0) base_cpu_processCpuTime_seconds - (T-1) base_cpu_processCpuTime_seconds) / (T0 - T-1) / (T0) base_cpu_availableProcessors)
```
Note - how does this differ from base_cpu_processCpuLoad_percent?

## Heap Utilization Percent
Grafana dashboard (Heap Usage Percentage): 
```
base_memory_usedHeap_bytes{instance=~`[[instance]]`} / base_memory_maxHeap_bytes{instance=~`[[instance]]`}
```
Built-in Liberty metric:
```
Gauge
vendor_memory_heapUtilization_percent
(T0) base_memory_usedHeap_bytes / (T0) base_memory_maxHeap_bytes
```

## REST Elapsed Time per Request
Grafana dashboard (Average Response Time per REST Request Over Last n min):
```
rate(base_REST_request_elapsedTime_seconds{instance=~`[[instance]]`}[[dur]]) / rate(base_REST_request_total{instance=~`[[instance]]`}[[dur]])
```
Built-in Liberty metric:
```
Gauge
vendor_REST_elapsedTime_per_request
EWMA(((T0) base_REST_request_elapsedTime_seconds - (T-1) base_REST_request_elapsedTime_seconds) / ((T0) base_REST_request_total - (T-1) base_REST_request_total))
```

## Servlet Elapsed Time per Request
Grafana dashboard (Average Response Time per Request Over Last n min):
```
rate(vendor_servlet_responseTime_total_seconds{instance=~`[[instance]]`}[[dur]])/rate(vendor_servlet_request_total{instance=~`[[instance]]`}[[dur]])
```
Built-in Liberty metric:
```
Gauge
vendor_servlet_elapsedTime_per_request
same as previous forumla, using vendor_servlet_responseTime_total_seconds, vendor_servlet_request_total
```

## Connection Pool in Use Time per Used Connection
Grafana dashboard (Average Time Usage per Connection Over Last n min):
```
rate(vendor_connectionpool_inUseTime_total_seconds{instance=~`[[instance]]`}[[dur]]) / rate(vendor_connectionpool_usedConnections_total{instance=~`[[instance]]`}[[dur]]) > 0
```
Built-in Liberty metric:
```
Gauge
vendor_connectionpool_inUseTime_per_usedConnection
same as previous forumla, using vendor_connectionpool_inUseTime_total_seconds, vendor_connectionpool_usedConnections_total
```

## Connection Pool Wait Time per Queued Request
Grafana dashboard (Average Wait Time per Queued Connection Request Over Last n min):
```
rate(vendor_connectionpool_waitTime_total_seconds{instance=~`[[instance]]`}[[dur]]) / rate(vendor_connectionpool_queuedRequests_total{instance=~`[[instance]]`}[[dur]]) > 0
```
Built-in Liberty metric:
```
Gauge
vendor_connectionpool_waitTime_per_queuedRequest
same as previous forumla, using vendor_connectionpool_waitTime_total_seconds, vendor_connectionpool_queuedRequests_total
```

## GC Time per GC Cycle
Grafana dashboard (Time per Garbage Collection Cycle n min moving average):
```
rate(base_gc_time_seconds{instance=~`[[instance]]`}[[dur]]) / rate(base_gc_total{instance=~`[[instance]]`}[[dur]])
```
Built-in Liberty metric:
```
Gauge
vendor_gc_time_per_gcCycle
same as previous forumla, using base_gc_time_seconds, base_gc_total
```

**Comment by abutch3r** [comment]
My notes from the UFO Socialization

As-is (Slide 7)
- [x] There is an item [2], but no [1]

To-Be (Slide 8)
- [x] I would add a statement that this will be applicable from 3.0 onwards, given 2.3 downwards is not being added to

Feature design (Slide 9/10)
- [x] Wording for Heap Util description given MaxHeap definitions being different between java and metrics definitions
- [x] Check the cpu_availableProcessors is processed as a float not an integer and make that clear in here
- [x] State why it is 5 minutes for EWMA
- [x] State how the 15 seconds relates to the EWMA 5 minutes - 20 samples take every 15 seconds for a 5 minute period
- [x] Include what happens with idle time handling when there are no requests
An additional slide might be needed for the last 3.

End user overview(Slide 12)
- [x] Call out Metrics 3-5.x
- [x] Add "Alerting" capabilities that can be achieved via the new metrics
- [x] Clarify the "percentage" values are between 0-1

Config (Slide 17)
- [x] Simplify server.xml to just be the feature section as the other sections are not relevant to the feature. the monitor filter section is sufficient for other metrics config configuration.
  - [x] State mpMetrics if the required feature and there is no additional toggle for computed metrics
- [x] Clarify what happens when the two required filter entries are not present - are the metrics still computed and not shown, or not computed
- [x] Do any of the new metrics only need one of the stated filter entries to be defined, or do all the new metrics need both. If both are needed for all it needs to be stated. If some don't need both then a table would be helpful to say Metric X needs filter F1, Metric Y needs filter F2 and Metric Z needs F1 and F2.
- [x] Callout that there is no configuration option for the 15 seconds between computation executions, that is set by OL.

Automated testing (Slide 24)
- [x] Clarify the `x.y` for the fats - especially if one fat covers more then one version

SVT (Slide 25)
- [x] Test for a non-Int number of cpus
- [x] Dashboard testing
	-  [x] AppFlow/CloudWatch

Performance (Slide 26)
- [x] Check the impact when computation occurs is tested

Platform/cloud considerations (Slide 27)
- [x] Callout will work with Instant-on

Serviceability (Slide 29)
- [x] Remove L3 line as those are typically no longer put in OL UFOs

Migration (Slide 31)
- [x] For this I would say either N/A as there is nothing the user needs to do to get the new metrics. Or state they need to add pre-existing filter values to the `<monitor>` block if they want these metrics

**Comment by pgunapal** [comment]
@abutch3r Thanks Alex, I updated the UFO with the suggested notes.

**Comment by pgunapal** [comment]
ID doc issue created: https://github.com/OpenLiberty/docs/issues/6977

**Comment by pgunapal** [comment]
@OpenLiberty/externals-approvers Can you please review the APIs/Externals approval for this feature? There are no API changes in this feature. It uses the APIs from the MpMetrics feature.

**Comment by pgunapal** [comment]
@OpenLiberty/performance-approvers Can you please review the Performance approval for this feature?

**Comment by pgunapal** [comment]
@OpenLiberty/svt-approvers Can you please review the SVT approval for this feature?

**Comment by pgunapal** [comment]
@OpenLiberty/serviceability-approvers - See the completed questions below. Please let me know if serviceability approval can be granted or if anything else is needed.

UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

Yes, the UFO has a slide for Serviceability which remains unchanged from the MpMetrics features (3.0, 4.0, 5.0, 5.1) that this feature supports. The slide also specifies the trace specification that can be enabled to get more diagnostic information to see if the computed metrics are being calculated properly. Also, provides the common problems users may encounter when using the MpMetrics feature.

Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).

a) What problem paths were tested and demonstrated?

The basic functionality was tested and demo'ed, when specific apps (Servlets, ConnectionPool, and REST) are loaded and hit, the corresponding computed metrics were shown in the /metrics output, in the correct format with correct names. The base computed metrics such as ProcessCPUUtilization, HeapUtilization, and GCTimePerCycle computed metrics were also demo'ed.

b) Who did you demo to?

Demo'ed to the Observability squad, as well as stakeholder teams, via visualizing the computed metrics in AWS CloudWatch dashboards.

c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

Yes

SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?

Daniel Guinan

b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?

Yes

Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.

WAS L2: CET
WAS L3: Liberty Log Analytics

Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

Yes, doc issue has been created to update the Metrics Reference list for the corresponding MpMetrics versions (https://github.com/OpenLiberty/docs/issues/6977)

**Comment by chirp1** [comment]
Doc issue from Prashanth at https://github.com/OpenLiberty/docs/issues/6977 Contains the info we need to document the feature. Doable for 23.0.0.11. Approving.

**Comment by pgunapal** [comment]
@OpenLiberty/demo-approvers Demo scheduled for EOI 23.21

**Comment by pgunapal** [comment]
Beta guards removed: https://github.com/OpenLiberty/open-liberty/pull/26662

**Comment by pgunapal** [comment]
All work completed. Delivered in 23.0.0.11.
